File: frontend/firebrat_app/lib/models/manifest.dart
class ManifestFigure {
  final String figureId;
  final String caption;
  final String imagePath;
  final int page;
  final int? width;
  final int? height;

  const ManifestFigure({
    required this.figureId,
    required this.caption,
    required this.imagePath,
    required this.page,
    this.width,
    this.height,
  });

  factory ManifestFigure.fromJson(Map<String, dynamic> json) => ManifestFigure(
        figureId: json['figure_id'] as String,
        caption: json['caption'] as String? ?? '',
        imagePath: json['image_path'] as String,
        page: json['page'] as int? ?? 0,
        width: json['width'] as int?,
        height: json['height'] as int?,
      );
}

class ManifestFormula {
  final String formulaId;
  final String latex;
  final String imagePath;
  final String spokenText;
  final bool visuallyEssential;
  final int page;

  const ManifestFormula({
    required this.formulaId,
    required this.latex,
    required this.imagePath,
    required this.spokenText,
    required this.visuallyEssential,
    required this.page,
  });

  factory ManifestFormula.fromJson(Map<String, dynamic> json) => ManifestFormula(
        formulaId: json['formula_id'] as String,
        latex: json['latex'] as String? ?? '',
        imagePath: json['image_path'] as String,
        spokenText: json['spoken_text'] as String? ?? '',
        visuallyEssential: json['visually_essential'] as bool? ?? false,
        page: json['page'] as int? ?? 0,
      );
}

class ManifestTable {
  final String tableId;
  final String caption;
  final String imagePath;
  final int page;

  const ManifestTable({
    required this.tableId,
    required this.caption,
    required this.imagePath,
    required this.page,
  });

  factory ManifestTable.fromJson(Map<String, dynamic> json) => ManifestTable(
        tableId: json['table_id'] as String,
        caption: json['caption'] as String? ?? '',
        imagePath: json['image_path'] as String,
        page: json['page'] as int? ?? 0,
      );
}

class ManifestSection {
  final String sectionId;
  final int? chapter;
  final int order;
  final String title;
  final String audioPath;
  final String segmentsPath;
  final int durationMs;
  final List<String> figureRefs;
  final List<String> formulaRefs;
  final List<String> tableRefs;
  final bool needsReview;

  const ManifestSection({
    required this.sectionId,
    this.chapter,
    required this.order,
    required this.title,
    required this.audioPath,
    required this.segmentsPath,
    required this.durationMs,
    required this.figureRefs,
    required this.formulaRefs,
    required this.tableRefs,
    required this.needsReview,
  });

  factory ManifestSection.fromJson(Map<String, dynamic> json) => ManifestSection(
        sectionId: json['section_id'] as String,
        chapter: json['chapter'] as int?,
        order: json['order'] as int,
        title: json['title'] as String,
        audioPath: json['audio_path'] as String,
        segmentsPath: json['segments_path'] as String,
        durationMs: json['duration_ms'] as int? ?? 0,
        figureRefs: (json['figure_refs'] as List? ?? []).cast<String>(),
        formulaRefs: (json['formula_refs'] as List? ?? []).cast<String>(),
        tableRefs: (json['table_refs'] as List? ?? []).cast<String>(),
        needsReview: json['needs_review'] as bool? ?? false,
      );
}

class Manifest {
  final String bookId;
  final String title;
  final String author;
  final String sourcePdf;
  final String generatedAt;
  final int totalDurationMs;
  final List<ManifestSection> sections;
  final List<ManifestFigure> figures;
  final List<ManifestFormula> formulas;
  final List<ManifestTable> tables;

  const Manifest({
    required this.bookId,
    required this.title,
    required this.author,
    required this.sourcePdf,
    required this.generatedAt,
    required this.totalDurationMs,
    required this.sections,
    required this.figures,
    required this.formulas,
    required this.tables,
  });

  factory Manifest.fromJson(Map<String, dynamic> json) => Manifest(
        bookId: json['book_id'] as String,
        title: json['title'] as String,
        author: json['author'] as String? ?? '',
        sourcePdf: json['source_pdf'] as String? ?? '',
        generatedAt: json['generated_at'] as String? ?? '',
        totalDurationMs: json['total_duration_ms'] as int? ?? 0,
        sections: (json['sections'] as List? ?? [])
            .map((e) => ManifestSection.fromJson(e as Map<String, dynamic>))
            .toList(),
        figures: (json['figures'] as List? ?? [])
            .map((e) => ManifestFigure.fromJson(e as Map<String, dynamic>))
            .toList(),
        formulas: (json['formulas'] as List? ?? [])
            .map((e) => ManifestFormula.fromJson(e as Map<String, dynamic>))
            .toList(),
        tables: (json['tables'] as List? ?? [])
            .map((e) => ManifestTable.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  ManifestFigure? figureById(String id) {
    for (final f in figures) {
      if (f.figureId == id) return f;
    }
    return null;
  }

  ManifestFormula? formulaById(String id) {
    for (final f in formulas) {
      if (f.formulaId == id) return f;
    }
    return null;
  }

  ManifestTable? tableById(String id) {
    for (final t in tables) {
      if (t.tableId == id) return t;
    }
    return null;
  }
}

File: frontend/FirebratNative/FirebratNative/Models/Manifest.swift
import Foundation

/// Mirror of backend/firebrat/pipeline/schema.py manifest models, with the same
/// tolerant defaults as frontend/firebrat_app/lib/models/manifest.dart:
/// missing optional fields fall back to "" / 0 / [] instead of failing the
/// whole manifest decode.

// MARK: - Assets

public struct ManifestFigure: Codable, Equatable, Sendable {
    public var figureId: String
    public var caption: String
    public var imagePath: String
    public var page: Int
    public var width: Int?
    public var height: Int?

    enum CodingKeys: String, CodingKey {
        case figureId = "figure_id", caption, imagePath = "image_path"
        case page, width, height
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        figureId = try c.decodeIfPresent(String.self, forKey: .figureId) ?? ""
        caption = try c.decodeIfPresent(String.self, forKey: .caption) ?? ""
        imagePath = try c.decodeIfPresent(String.self, forKey: .imagePath) ?? ""
        page = try c.decodeIfPresent(Int.self, forKey: .page) ?? 0
        width = try c.decodeIfPresent(Int.self, forKey: .width)
        height = try c.decodeIfPresent(Int.self, forKey: .height)
    }
}

public struct ManifestFormula: Codable, Equatable, Sendable {
    public var formulaId: String
    public var latex: String
    public var imagePath: String
    public var spokenText: String
    public var visuallyEssential: Bool
    public var page: Int

    enum CodingKeys: String, CodingKey {
        case formulaId = "formula_id", latex, imagePath = "image_path"
        case spokenText = "spoken_text", visuallyEssential = "visually_essential", page
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        formulaId = try c.decodeIfPresent(String.self, forKey: .formulaId) ?? ""
        latex = try c.decodeIfPresent(String.self, forKey: .latex) ?? ""
        imagePath = try c.decodeIfPresent(String.self, forKey: .imagePath) ?? ""
        spokenText = try c.decodeIfPresent(String.self, forKey: .spokenText) ?? ""
        visuallyEssential = try c.decodeIfPresent(Bool.self, forKey: .visuallyEssential) ?? false
        page = try c.decodeIfPresent(Int.self, forKey: .page) ?? 0
    }
}

public struct ManifestTable: Codable, Equatable, Sendable {
    public var tableId: String
    public var caption: String
    public var imagePath: String
    public var page: Int

    enum CodingKeys: String, CodingKey {
        case tableId = "table_id", caption, imagePath = "image_path", page
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        tableId = try c.decodeIfPresent(String.self, forKey: .tableId) ?? ""
        caption = try c.decodeIfPresent(String.self, forKey: .caption) ?? ""
        imagePath = try c.decodeIfPresent(String.self, forKey: .imagePath) ?? ""
        page = try c.decodeIfPresent(Int.self, forKey: .page) ?? 0
    }
}

// MARK: - Sections

public struct ManifestSection: Codable, Equatable, Sendable {
    public var sectionId: String
    public var chapter: Int?
    public var order: Int
    public var title: String
    public var audioPath: String
    public var segmentsPath: String
    public var durationMs: Int
    public var figureRefs: [String]
    public var formulaRefs: [String]
    public var tableRefs: [String]
    public var needsReview: Bool

    enum CodingKeys: String, CodingKey {
        case sectionId = "section_id", chapter, order, title
        case audioPath = "audio_path", segmentsPath = "segments_path"
        case durationMs = "duration_ms"
        case figureRefs = "figure_refs", formulaRefs = "formula_refs", tableRefs = "table_refs"
        case needsReview = "needs_review"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        sectionId = try c.decodeIfPresent(String.self, forKey: .sectionId) ?? ""
        chapter = try c.decodeIfPresent(Int.self, forKey: .chapter)
        order = try c.decodeIfPresent(Int.self, forKey: .order) ?? 0
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? ""
        audioPath = try c.decodeIfPresent(String.self, forKey: .audioPath) ?? ""
        segmentsPath = try c.decodeIfPresent(String.self, forKey: .segmentsPath) ?? ""
        durationMs = try c.decodeIfPresent(Int.self, forKey: .durationMs) ?? 0
        figureRefs = try c.decodeIfPresent([String].self, forKey: .figureRefs) ?? []
        formulaRefs = try c.decodeIfPresent([String].self, forKey: .formulaRefs) ?? []
        tableRefs = try c.decodeIfPresent([String].self, forKey: .tableRefs) ?? []
        needsReview = try c.decodeIfPresent(Bool.self, forKey: .needsReview) ?? false
    }
}

// MARK: - Manifest

public struct Manifest: Codable, Equatable, Sendable {
    public var bookId: String
    public var title: String
    public var author: String
    public var sourcePdf: String
    public var generatedAt: String
    public var totalDurationMs: Int
    public var sections: [ManifestSection]
    public var figures: [ManifestFigure]
    public var formulas: [ManifestFormula]
    public var tables: [ManifestTable]

    enum CodingKeys: String, CodingKey {
        case bookId = "book_id", title, author, sourcePdf = "source_pdf"
        case generatedAt = "generated_at", totalDurationMs = "total_duration_ms"
        case sections, figures, formulas, tables
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        bookId = try c.decodeIfPresent(String.self, forKey: .bookId) ?? ""
        title = try c.decodeIfPresent(String.self, forKey: .title) ?? ""
        author = try c.decodeIfPresent(String.self, forKey: .author) ?? ""
        sourcePdf = try c.decodeIfPresent(String.self, forKey: .sourcePdf) ?? ""
        generatedAt = try c.decodeIfPresent(String.self, forKey: .generatedAt) ?? ""
        totalDurationMs = try c.decodeIfPresent(Int.self, forKey: .totalDurationMs) ?? 0
        sections = try c.decodeIfPresent([ManifestSection].self, forKey: .sections) ?? []
        figures = try c.decodeIfPresent([ManifestFigure].self, forKey: .figures) ?? []
        formulas = try c.decodeIfPresent([ManifestFormula].self, forKey: .formulas) ?? []
        tables = try c.decodeIfPresent([ManifestTable].self, forKey: .tables) ?? []
    }

    public static func load(from url: URL) throws -> Manifest {
        try JSONDecoder().decode(Manifest.self, from: Data(contentsOf: url))
    }

    public func figure(id: String) -> ManifestFigure? {
        figures.first { $0.figureId == id }
    }

    public func formula(id: String) -> ManifestFormula? {
        formulas.first { $0.formulaId == id }
    }

    public func table(id: String) -> ManifestTable? {
        tables.first { $0.tableId == id }
    }
}
